Add link provider for includes, closes #99

danyill opened the pull request:
**What I did:**

*    Used an IncludeProcessor to find the link text and reader line numbers for includes and then regex parsed the document to identify line numbers more precisely since the Asciidoctor reader is "imprecise" (optimised for speed, not accuracy).
* Created a line number alignment algorithm to process include line numbers from the IncludeProcessor and from regex parsing to "match up" the relevant lines.
* Abstracted the link sanitizer code previously only used in the preview as part of AsciidoctorWebViewConverter to be part of the normalizeLinks routine so that preview and DocumentLinks use the same implementation.

**Implementation limitations:**

Items within an `ifdef` or `ifeval` which aren't processed won't receive DocumentLinks because they're not handled by the include processor. But to avoid this would require quite sophisticated external logic in this extension and doesn't currently seem worthwhile.

**What I didn't do:**

Write a test. I'm not entirely sure how but I haven't spent any time on it. I have commented out the tests because it was causing the linter to be argumentative.
I can work on this if you would prefer to have this done prior to merging but I would like to see if we think this PR is heading in the right direction.

I've done some initial testing and the logic seems to work across a range of documents that were close at hand... but some more testing would be appreciated.

_NOTE:_ Originally this PR was part of #474 but this was trying to do inline links as well and this was not possible and I was unable to effectively rebase it so I've taken a diff of the final commit to the originating commit to generate this PR and rebased on the current master.

Mogztter commented on src/asciidoctorFindIncludeProcessor.ts:
```diff
@@ -0,0 +1,49 @@
+interface IncludeEntry {
+  name: string
+  position: number
+  length: string
+}
+
+interface IncludeItem {
+  [index: number]: IncludeEntry
+}
+
+interface IncludeItems extends Array<IncludeItem>{}
+
+let baseDocIncludes: IncludeItems[] = []
```
I think it would be better to use an array of `IncludeEntry` rather than an array of `Array<IncludeItem>`

```js
[
  { name: 'a.adoc', position: 5, length: 6 },
  { name: 'github.adoc', position: 10, length: 11 },
  { name: 'windows.adoc', position: 16, length: 12 },
  { name: 'b.adoc', position: 27, length: 6 },
  { name: 'a.adoc', position: 31, length: 6 },
  { name: 'c.adoc', position: 45, length: 6 }
]
```

danyill replied:
Yes, it is better. This was my hacky way of getting an index position easily. I will improve.

Mogztter commented:
> Items within an ifdef or ifeval which aren't processed won't receive DocumentLinks because they're not handled by the include processor. But to avoid this would require quite sophisticated external logic in this extension and doesn't currently seem worthwhile.

I can think of an idea to workaround that but we should discuss it in a follow-up issue/pull request.
Another limitation is that attributes won't be resolved if they are defined in an include.

**main.adoc**
```
include::attrs.adoc[]

include::{operating-system-name}.adoc[]
```

**attrs.adoc**
```
:operating-system: macos
```

Again we should address this limitation in another issue/pull request.

Mogztter commented:
> Write a test. I'm not entirely sure how but I haven't spent any time on it. I have commented out the tests because it was causing the linter to be argumentative.

On my machine the `npm run lint` task runs fine if I uncomment `documentLinkProvider.test.ts`

Mogztter commented on src/asciidocParser.ts:
```diff
@@ -2,42 +2,37 @@ import * as vscode from 'vscode'
-
-  public getAttribute (name: string) {
-    return (this.document == null) ? null : this.document.getAttribute(name)
+  constructor (extensionUri: vscode.Uri, private errorCollection: vscode.DiagnosticCollection = null) {
```
This class desperately needs to be rewritten.
We should have dissociate the `load` function from the `convert` function and avoid using mutable class variables.

The `load` and `convert` functions should be pure.

danyill replied:
I'm happy to look at this, perhaps we can discuss on the issue.

Mogztter commented on src/image-paste.ts:
```diff
@@ -327,10 +327,12 @@ export namespace Import {
-        dir = adoc.document.getAttribute('imagesdir')
+        const textDocument = vscode.window.activeTextEditor.document
+        const extensionUri = vscode.Uri.file('') // won't be used anyway... needs refactoring!
+        const {document} = await new AsciidocParser(extensionUri).parseText(textDocument.getText(), textDocument)
```
Since we don't need to convert the document, we don't need to define a valid `extensionUri`.

Mogztter commented:
@danyill tests are running fine now, I had to remove `vscode.extensions.getExtension('asciidoctor.asciidoctor-vscode')` (which is, I believe, an anti-pattern anyway).

I'm already much more confident now that we have a few automated tests running. The code coverage is pretty low but that's a start.

Since I made quite a few changes, feel free to manually test to make sure that I didn't break existing features.
I'm ready to merge when you are ready.

In a following pull request I would like to rewrite/dissociate/simplify `AsciidocEngine` and `AsciidocParser`.

danyill commented:
>  Another limitation is that attributes won't be resolved if they are defined in an include.

Really good point and it didn't occur to me at all.

I think we could perhaps do something like what was suggested on Zulip by Dan here: https://asciidoctor.zulipchat.com/#narrow/stream/279642-users/topic/Adding.20'optional'.20include.3A.3A.20option.20via.20extension/near/261486251

I filed: https://github.com/asciidoctor/asciidoctor-vscode/issues/496

danyill commented:
Thank you for this, I am delighted to have some tests which we can build on.

Thank you also for your review comments, I have changed the way IncludeEntries are handled to use a better object definition. I appreciate the tidying of the code base as you make other changes. There is much tidying possible.

I force pushed this because this did not resolve #99, but rather #498 and I wanted to fix the commit history. I added a CHANGELOG entry, hoping to get this into the next release.

If you are happy feel free to merge.

Mogztter commented:
It looks good to me, nice work 👍
